backend/invertedIndex.js

```diff
@@ -18,21 +30,42 @@ const buildInvertedIndex = async () => {
   const invertedIndex = {};
 
   businesses.forEach((business) => {
+    // Tokenize and normalize the business name, category, overview, and business type
     const keywords = [
       ...tokenizeAndNormalize(business.name),
       ...tokenizeAndNormalize(business.category),
       ...tokenizeAndNormalize(business.overview),
+      ...tokenizeAndNormalize(business.businessType),
     ];
 
     keywords.forEach((keyword) => {
-      if (!invertedIndex[keyword]) {
-        invertedIndex[keyword] = [];
+      // Create an entry in the inverted index for each prefix of the keyword
+      for (let i = 1; i <= keyword.length; i++) {
+        const prefix = keyword.slice(0, i);
+        if (!invertedIndex[prefix]) {
+          invertedIndex[prefix] = [];
+        }
+        invertedIndex[prefix].push(business);
       }
-      invertedIndex[keyword].push(business);
     });
   });
 
   return invertedIndex;
 };
 
-module.exports = { buildInvertedIndex };
+// Global variable to store the inverted index
+let invertedIndex;
+
+const initializeInvertedIndex = async () => {
+  invertedIndex = await buildInvertedIndex();
+};
+
+// Initialize the inverted index when the module is loaded
+initializeInvertedIndex();
```

Comment:
hm, why do we call `buildInvertedIndex()` when the module is loaded? It seems like we call the same function in `searchBusinesses()`.  Does the index need to be built once at startup, or does it need to happen on each request that accesses the index (such that the index is up-to-date)

Reply:
We call buildInvertedIndex() when the module is loaded to ensure that the inverted index is built once at startup. However, in searchBusinesses(), we also call buildInvertedIndex() to ensure the index is up-to-date.